# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 1303 Беззубов Даниил. Вариант №2.

## Цель работы

Изучить понятие аддитивной цепочки, познакомиться с алгоритмами построения аддитивных цепочек и реализовать их, получить аддитивные цепочки для различных чисел, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

### Понятие аддитивной цепочки

_Определение._<br>
Аддитивной цепочкой для натурального $n$ называется последовательность натуральных чисел<br>
$$ 1 = a_0, a_1, a_2,\dots,a_r = n, $$
где каждый элемент последовательности равен сумме каких-либо двух предыдущих элементов:
$$ a_i = a_j + a_k,\quad k \leqslant j < i,\quad i = 1, 2, \dots, r. $$

_Связь аддитивных цепочек с методами эффективного возведения мономов в степень._<br>
<ul>
    <li>$l(n) = r$ - наименьшая длина $r$, для которой существует аддитивная цепочка для значения $n$(функция от $n$);</li>
    <li>Неравенство для метода множителей:</li>
        $$ l(m \cdot n) \leqslant l(m) + l(n).$$
    <li>Неравенство $m$-арного метода($m = 2^k, n = \sum\limits_{j=0}^td_jm^{t-j}$):</li>
        $$ l(n) \leqslant m \cdot 2 + (k + 1)^t.$$
    <li>Неравенство для бинарного метода "SX":</li>
        $$ l(n) \leqslant \lambda(n) + \nu(n) - 1. $$
</ul>

_Свойства аддитивных цепочек._<br>
_Шаг аддитивной цепочки._<br>
<ul>
    <li>Полагаем, что все аддитивные цепочки возрастающие:</li>
        $$ 1 = a_0 < a_1 < a_2 < \dots < a_r = n. $$
    <li>Если два числа из $\{a_i\}$ одинаковые, то одно из них может быть опущено;</li>
    <li>Пара $(j, k), 0 \leqslant k \leqslant j < i$ - называется шагом;</li>
    <li>Если существует более чем одна пара $(j, k)$, полагаем, что $j$ - наибольшее из возможных.</li>
    </ul>
    
_Виды шагов._
<ul>
    <li>Удвоение:</li>
        $ j = k = i - 1 $
    <li>Звёздный:</li>
        $ j = i - 1 $
    <li>Малый:</li>
        $ \lambda(a_i) = \lambda(a_{i-1}) $
    </ul>
    
_Свойства шагов._
<ul>
    <li>Первый шаг всегда удвоение;</li>
    <li>Удвоение - звёздный шаг, но никогда не малый шаг;</li>
    <li>За удвоением всегда следует звёздный шаг;</li>
    <li>Если $i$-ый шаг не малый, то $(i+1)$ шаг либо малый, либо звёздный, либо и тот, и другой;</li>
    <li>Если $(i+1)$ шаг ни звездный, ни малый, то $i$-ый шаг должен быть малым.</li>
    </ul>
    
### Алгоритм Яо

<ul>
    <li>Алгоритм Яо имеет ту же вычислительную сложность, что и алгоритм Брауэра.</li>
    <li>Начало также схоже: задается некоторое целое $k \geqslant 2$ и число $n$ раскладывается в $2^k$-ой системе счисления.</li>
    $$ n = \sum\limits_{i=0}^ja_i2^{ik},\quad a_{j} \neq 0$$
    <li>Введем функцию d:</li>
    $$ d(z) = \sum\limits_{i:a_i=z}2^{ik}$$
    </ul>
    
_Ход алгоритма._<br>
<ol>
    <li>Задаем базовую последовательность $1, 2, 4, \dots, 2^{\lambda(n)}$;</li>
    <li>вычисление значения $d(z)$ для всех $z \in \{1, 2, 3, \dots, 2^k-1\}$, причем $d(z) \neq 0$;</li>
    <li>Вычисление $z\cdot d(z)$ для $\forall z$;</li>
    <li>В конечном итоге:</li>
    $$ n = \sum\limits_{z=1}^{2^k-1}zd(z).$$
</ol>

### Звёздные цепочки.
_Определение._<br>
Звездной цепочкой называется аддитивная цепочка, включающая в себя только звёздные шаги.<br>
<ul>
    <li>Длина звездной цепочки $l^*(n)$.</li>
    <li>Шаг звёздной цепочки $a_i = a_{i-1} + a_k \forall k < i$.</li>
    <li>Очевидно, что $l(n) \leqslant l^*(n)$.</li>
    </ul>
    
### Вектор индексов
_Определение._<br>
Пусть дана звездная цепочка длины $m-1$ вида $1 = a_1, a_2, \dots, a_m$. Для каждой звездной цепочки существует вектор индексов $r = \{r_1, r_2, \dots, r\}$ длины $m-1$ такой, что
$$ r_i = \{z:1\leqslant z\leqslant i\},\quad a_i = a_{i-1} + a_{r_{i-1}},\quad 2\leqslant i\leqslant m-1$$
<ul>
    <li>Первый элемент вектора $r_1$ - всегда $1$, второй $r_2 = \{1, 2\}$ и так далее. Последний элемент $1, \dots, m_1$.</li>
    <li>Наибольшая звездная цепочка с соответствующим ей вектором индексов имеют вид:</li>
    $a = \{1, 2, 4, \dots, 2^{m-1}\}$<br>
    $r = \{1, 2, 3, \dots, m-1\}$<br>
    <li>Наименьшая звездная цепочка с соответствующим ей вектором индексов имеют вид:</li>
    $a = \{1, 2, 4, \dots, m\}$<br>
    $r = \{1, 1, 1, \dots, 1\}$<br>
    <li>#$ A^*(m-1) = (m-1)! $</li>
    </ul>
    
### Алгоритм дробления вектора индексов

С помощью данного алгоритма можно найти звездную цепочку для числа $n$. Укажем шаги работы алгоритма.<br>
1. Запускаем внешний цикл по длинам $\underline{l}(n) \le m \le \overline{l}(n)$, выбираем индекс дроблений $q = \frac{m}{2}$
2. Внутренний цикл - перебор всех фиксированных частей. На каждом шагу вычисляем $a_{min}, a_{max}$ и строим цепочку
    + Если $a_m = n$, то задача решена
    + Если $n \notin [a_{min}, a_{max}]$, то переходим к следующему набору ${r_1, r_2, \dots, r_q}$
    + Если $n \in [a_{min}, a_{max}]$, то организуем внутренний цикл перебора меняющейся части
        * Если $a_m = a_n$, то задача решена
        * Иначе переходим к следующему (по введенной упорядоченности) фиксированной части $\{\rho_{q+1}, \rho_{q+2}, \dots, \rho_m\}$
3. Если все наборы фиксированной длины исчерпаны, то увеличиваем их длину во внешнем цикле

### Гипотеза Шольца-Брауэра

Гипотеза заключается в следующем:
$$l(2^n - 1) \le l(n) + n - 1$$
При этом:
* Гипотеза доказана для звездных цепочек
* Гипотеза справедлива для всех $n < 578469$
* Равенство выполняется для всех $n \le 64$

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы


### Реализация алгоритма Яо

Реализуем алгоритм Яо. Для этого напишем вспомогательные для алгоритма функции convert_base (переводит число n в систему счисления b), get_d (вычисляет значение фунции d(z)), sx() (реализация метода SX) а также функцию Yao_method, собирающую аддитивную цепочку в переменную yao_chain и возвращающую её.

In [93]:
pip install prettytable

You should consider upgrading via the '/opt/sagemath-9.3/local/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [23]:
def convert_base(n, b):
    if n == 0:
        return [n]
    res = []
    while n > 0:
        res.append(int(n % b))
        n //= b
    return res

def get_d(num, z, k):
    return sum(2^(i*k) for i, elem in enumerate(num) if elem == z)

def sx(power):
    if power < 1:
        return [1]
    bin_power = str(bin(power))[3:]
    sx_representation = bin_power.replace('1','SX').replace('0', 'S')
    result = 1
    return_list = [result]
    for i, elem in enumerate(sx_representation):
        if elem == 'X':
            result += 1
        else:
            result *= 2
        return_list.append(result)
    return return_list

def Yao_method(n, k):
    _lambda = floor(log(n, 2))
    yao_chain = [2 ^ i for i in range(_lambda + 1)]
    num = convert_base(n, 2 ^ k)
    d_arr = []
    for z in range(1, 2^k):
        d = get_d(num, z, k)
        d_arr.append(d)
        if d != 0:
            yao_chain += [elem*d_arr[z-1] for elem in sx(z)]
    yao_chain.append(n)
    return sorted(set(yao_chain))

Yao_method(5,2)

[1, 2, 4, 5]

Продемонстрируем работу алгоритма для $n = 15, 57, 67, 107, 167$ и $k = 2, 3, 4, 5, 6$, при этом будем варьировать параметр k, сравнивать длину получившейся цепочки с длиной минимальной цепочки.

In [24]:
from prettytable import PrettyTable
from prettytable import MARKDOWN

def test_Yao():
#=====Пресет таблицы============== 
    x = PrettyTable()
#     x.set_style(MARKDOWN)
    x.field_names = ["k", "n", "Yao", "Yao len", "Shortest len"]
    
#=======тестовые данные===========
    k = [*range(2,7)]
    test_data = [15, 57, 65, 107, 167]
    shortest_chains = {15:[1,2,3,6,12,15], 57:[1,2,3,6,7,14,28,56,57], 65:[1,2,4,8,16,32,64,65], 107:[1,2,3,6,12,13,26,52,104,107], 167:[1,2,3,5,10,20,40,80,83,166,167]}
#=================================   

    for elem in test_data:
        for i in k:
            yao_result = Yao_method(elem, i)
            length = len(yao_result)
            x.add_row([i, elem, yao_result, length, len(shortest_chains[elem])])
    
    print(x)
    
test_Yao()

+---+-----+-------------------------------------------------------+---------+--------------+
| k |  n  |                          Yao                          | Yao len | Shortest len |
+---+-----+-------------------------------------------------------+---------+--------------+
| 2 |  15 |                [1, 2, 4, 5, 8, 10, 15]                |    7    |      6       |
| 3 |  15 |               [1, 2, 3, 4, 6, 7, 8, 15]               |    8    |      6       |
| 4 |  15 |             [1, 2, 3, 4, 6, 7, 8, 14, 15]             |    9    |      6       |
| 5 |  15 |             [1, 2, 3, 4, 6, 7, 8, 14, 15]             |    9    |      6       |
| 6 |  15 |             [1, 2, 3, 4, 6, 7, 8, 14, 15]             |    9    |      6       |
| 2 |  57 |              [1, 2, 4, 8, 16, 32, 48, 57]             |    8    |      9       |
| 3 |  57 |          [1, 2, 4, 8, 16, 24, 32, 48, 56, 57]         |    10   |      9       |
| 4 |  57 |            [1, 2, 4, 8, 9, 16, 32, 48, 57]            |   

Внесем полученные результаты в таблицу:

| k |  n  |                          Yao                          |               Shortest chain               | Yao len | Shortest len |
|:-:|:---:|:-----------------------------------------------------:|:------------------------------------------:|:-------:|:------------:|
| 2 |  15 |                [1, 2, 4, 5, 8, 10, 15]                |            [1, 2, 3, 6, 12, 15]            |    7    |      6       |
| 3 |  15 |               [1, 2, 3, 4, 6, 7, 8, 15]               |            [1, 2, 3, 6, 12, 15]            |    8    |      6       |
| 4 |  15 |             [1, 2, 3, 4, 6, 7, 8, 14, 15]             |            [1, 2, 3, 6, 12, 15]            |    9    |      6       |
| 5 |  15 |             [1, 2, 3, 4, 6, 7, 8, 14, 15]             |            [1, 2, 3, 6, 12, 15]            |    9    |      6       |
| 6 |  15 |             [1, 2, 3, 4, 6, 7, 8, 14, 15]             |            [1, 2, 3, 6, 12, 15]            |    9    |      6       |
| 2 |  57 |              [1, 2, 4, 8, 16, 32, 48, 57]             |      [1, 2, 3, 6, 7, 14, 28, 56, 57]       |    8    |      9       |
| 3 |  57 |          [1, 2, 4, 8, 16, 24, 32, 48, 56, 57]         |      [1, 2, 3, 6, 7, 14, 28, 56, 57]       |    10   |      9       |
| 4 |  57 |            [1, 2, 4, 8, 9, 16, 32, 48, 57]            |      [1, 2, 3, 6, 7, 14, 28, 56, 57]       |    9    |      9       |
| 5 |  57 |       [1, 2, 3, 4, 6, 8, 12, 16, 24, 25, 32, 57]      |      [1, 2, 3, 6, 7, 14, 28, 56, 57]       |    12   |      9       |
| 6 |  57 |     [1, 2, 3, 4, 6, 7, 8, 14, 16, 28, 32, 56, 57]     |      [1, 2, 3, 6, 7, 14, 28, 56, 57]       |    13   |      9       |
| 2 |  65 |              [1, 2, 4, 8, 16, 32, 64, 65]             |        [1, 2, 4, 8, 16, 32, 64, 65]        |    8    |      8       |
| 3 |  65 |              [1, 2, 4, 8, 16, 32, 64, 65]             |        [1, 2, 4, 8, 16, 32, 64, 65]        |    8    |      8       |
| 4 |  65 |              [1, 2, 4, 8, 16, 32, 64, 65]             |        [1, 2, 4, 8, 16, 32, 64, 65]        |    8    |      8       |
| 5 |  65 |              [1, 2, 4, 8, 16, 32, 64, 65]             |        [1, 2, 4, 8, 16, 32, 64, 65]        |    8    |      8       |
| 6 |  65 |              [1, 2, 4, 8, 16, 32, 64, 65]             |        [1, 2, 4, 8, 16, 32, 64, 65]        |    8    |      8       |
| 2 | 107 |        [1, 2, 3, 4, 8, 16, 20, 32, 40, 64, 107]       |   [1, 2, 3, 6, 12, 13, 26, 52, 104, 107]   |    11   |      10      |
| 3 | 107 |          [1, 2, 3, 4, 8, 16, 32, 40, 64, 107]         |   [1, 2, 3, 6, 12, 13, 26, 52, 104, 107]   |    10   |      10      |
| 4 | 107 |    [1, 2, 4, 5, 8, 10, 11, 16, 32, 48, 64, 96, 107]   |   [1, 2, 3, 6, 12, 13, 26, 52, 104, 107]   |    13   |      10      |
| 5 | 107 |      [1, 2, 4, 5, 8, 10, 11, 16, 32, 64, 96, 107]     |   [1, 2, 3, 6, 12, 13, 26, 52, 104, 107]   |    12   |      10      |
| 6 | 107 |  [1, 2, 4, 5, 8, 10, 16, 20, 21, 32, 42, 43, 64, 107] |   [1, 2, 3, 6, 12, 13, 26, 52, 104, 107]   |    14   |      10      |
| 2 | 167 |     [1, 2, 3, 4, 8, 16, 32, 64, 80, 128, 160, 167]    | [1, 2, 3, 5, 10, 20, 40, 80, 83, 166, 167] |    12   |      11      |
| 3 | 167 |      [1, 2, 3, 4, 6, 7, 8, 16, 32, 64, 128, 167]      | [1, 2, 3, 5, 10, 20, 40, 80, 83, 166, 167] |    12   |      11      |
| 4 | 167 |  [1, 2, 3, 4, 6, 7, 8, 16, 32, 64, 80, 128, 160, 167] | [1, 2, 3, 5, 10, 20, 40, 80, 83, 166, 167] |    14   |      11      |
| 5 | 167 |    [1, 2, 3, 4, 6, 7, 8, 16, 32, 64, 128, 160, 167]   | [1, 2, 3, 5, 10, 20, 40, 80, 83, 166, 167] |    13   |      11      |
| 6 | 167 | [1, 2, 4, 8, 9, 16, 18, 19, 32, 38, 39, 64, 128, 167] | [1, 2, 3, 5, 10, 20, 40, 80, 83, 166, 167] |    14   |11      |

### Вывод

Реализован алгоритм Яо построения аддитивной цепочки. Полученный алгоритм протестирован на нескольких значениях, полученные результаты внесены в таблицу. Основываясь на полученных данных, можно сделать вывод, что Алгоритм Яо не всегда дает оптимальную длину цепочки - это зависит, как от самого числа, так и от параметра k, при этом данный алгоритм можно считать эффективным благодаря скорости его работы.

## Реализация алгоритма дробления вектора индексов



Реализуем алгоритм дробления вектора индексов согласно теоретическим положениям. Для этого напишем вспомогательные для алгоритма функции build_chain (возвращает аддитивную цепочку по вектору индексов), next_index_vector (возвращает следующий относительно данного вектор индексов), index_vector_fraction (является реализацией алгоритма дробления вектора индексов).

In [25]:
import time


def build_chain(vec) -> list :
    chain = [1]
    for i in vec:
        chain.append(chain[-1] + chain[i-1])
    return chain

def next_index_vector(vec, q=0) -> list:
    if vec == [1]*len(vec): return []
    for i in range(len(vec)-1, -1, -1):
        if vec[i] == 1:
            vec[i] = i + q + 1
        else:
            vec[i] -= 1
            break
    return vec

def index_vector_fraction(n) -> list:
    
    if n == 1: return [1]
    if n <= 1: return []
    
    lower_border = ceil(log(n, 2))
    upper_border = floor(log(n, 2)) + bin(n).count('1')
    
    for m in range(lower_border, upper_border+1):
        q = m // 2 + 1
        index_vec = [*range(1, m+1)]
        
        mutable_ind = index_vec[:q]
        immutable_ind = index_vec[q:]
        
        while len(mutable_ind):
            chain = build_chain(mutable_ind+immutable_ind)
            
            if chain[-1] == n:
                return chain
            
            a_min = chain[q] + m - q
            a_max = chain[q] * (2^(m - q))
            
            if n > a_min and n < a_max :
                while len(immutable_ind):
                    chain = build_chain(mutable_ind+immutable_ind)
                    if chain[-1] == n:
                        return chain
                    immutable_ind = next_index_vector(immutable_ind, q)
                mutable_ind = next_index_vector(mutable_ind)
                immutable_ind = index_vec[q:]
            else:
                mutable_ind = next_index_vector(mutable_ind)
    return chain
                    

Продемонстрируем работу на пяти $n > 1000$, а именно $n = 1152, 1600, 1028, 1024, 1288$. Данные примеры подобраны таким образом, чтобы время расчетов было небольшим

In [30]:
from prettytable import PrettyTable
from prettytable import MARKDOWN

def test_ind_fract_alg():
    safe_test_array = [1152, 1028, 1344, 1024, 1288]
    
    x = PrettyTable()
    x.set_style(MARKDOWN)
    x.field_names = ["n", "Additive chain", "Time"]
    
    for elem in safe_test_array:
        t = time.time()
        index_vector = index_vector_fraction(elem)
        t = time.time() - t
        x.add_row([elem, index_vector, t])
        
    print(x)

test_ind_fract_alg()

|  n   |                       Additive chain                      |          Time          |
|:----:|:---------------------------------------------------------:|:----------------------:|
| 1152 |    [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152]    | 0.0005412101745605469  |
| 1028 |    [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1028]    | 0.0003974437713623047  |
| 1344 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1344] |   2.7421865463256836   |
| 1024 |       [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]       | 0.00011539459228515625 |
| 1288 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1288] |   2.741096019744873    |


Оформим результат в виде таблицы:

|  n   |                       Additive chain                      |          Time          |
|:----:|:---------------------------------------------------------:|:----------------------:|
| 1152 |    [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152]    | 0.0004911422729492188  |
| 1028 |    [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1028]    | 0.00036787986755371094 |
| 1344 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1344] |   2.747333288192749    |
| 1024 |       [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]       | 0.00010895729064941406 |
| 1288 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1288] |   2.733215570449829    |

### Вывод

Время выполнения расчетов в данных примерах не является репрезентативным, т.к. они подобраны специально для того, чтобы время рассчета было невелико. В общем случае данный алгоритм более ресурсозатратный, чем реализованный раннее алгоритм Яо, но при этом главным преимуществом алгоритма дробления вектора индексов является то, что он гарантированно дает звездную цепочку наименьшей длины.

## Проверка гипотезы Шольца-Брауэра

Проверим гипотезу Шольца-Бауэра для $n <= 12$. Т.к. $ \forall n \le 64$ известно, что $l(2^n - 1) = l(n) + n - 1$, воспользуемся данным фактом для оптимизации вычислений. Модифицируем алгоритм дробления вектора индексов, перебирая только длины, равные $l(n) + n - 1$

In [34]:
def index_vector_fraction_modify(n,m) -> list:
    
    if n == 1: return [1]
    if n <= 1: return []
    
    m = m-1
    q = m//2+1
    
    index_vec = [*range(1, m + 1)]
    
    mutable_ind = index_vec[:q]
    immutable_ind = index_vec[q:]

    while len(mutable_ind):
        chain = build_chain(mutable_ind+immutable_ind)

        if chain[-1] == n:
            return chain

        a_min = chain[q] + m - q
        a_max = chain[q] * (2^(m - q))

        if n > a_min and n < a_max :
            while len(immutable_ind):
                chain = build_chain(mutable_ind+immutable_ind)
                if chain[-1] == n:
                    return chain
                immutable_ind = next_index_vector(immutable_ind, q)
            mutable_ind = next_index_vector(mutable_ind)
            immutable_ind = index_vec[q:]
        else:
            mutable_ind = next_index_vector(mutable_ind)
    return chain

Реализуем функцию для проверки данной гипотезы для $n <= 12$

In [37]:
def check_hypothesis():
    
    test_set = [*range(1, 13)]
    
    x = PrettyTable()
    x.set_style(MARKDOWN)
    x.field_names = ["$n$", "$l(2^n-1)$", "$l(n) + n - 1$", "$Result$"]
    
    for elem in test_set:
        val_2 = len(index_vector_fraction(elem)) + elem -1
        val_1 = len(index_vector_fraction_modify(2^elem - 1, val_2))
        if val_1 > val_2:
            x.add_row([elem, val_1, val_2, "Не верно"])
            break
        else:
            x.add_row([elem, val_1, val_2, "Верно"])
    
    print(x)
    
check_hypothesis()

| $n$ | $l(2^n-1)$ | $l(n) + n - 1$ | $Result$ |
|:---:|:----------:|:--------------:|:--------:|
|  1  |     1      |       1        |  Верно   |
|  2  |     3      |       3        |  Верно   |
|  3  |     5      |       5        |  Верно   |
|  4  |     6      |       6        |  Верно   |
|  5  |     8      |       8        |  Верно   |
|  6  |     9      |       9        |  Верно   |
|  7  |     11     |       11       |  Верно   |
|  8  |     11     |       11       |  Верно   |
|  9  |     13     |       13       |  Верно   |
|  10 |     14     |       14       |  Верно   |
|  11 |     16     |       16       |  Верно   |
|  12 |     16     |       16       |  Верно   |


Оформим результат в виде таблицы:

|  $n$  |  $l(2^n-1)$  |  $l(n) + n - 1$  |  $Result$  |
|:---:|:----------:|:--------------:|:--------:|
|  $1$  |     $1$      |       $1$       |  $Верно$   |
|  $2$  |     $3$      |       $3$        |  $Верно$   |
|  $3$  |     $5$      |       $5$        |  $Верно$   |
|  $4$  |     $6$      |       $6$        |  $Верно$   |
|  $5$  |     $8$      |       $8$       |  $Верно$   |
|  $6$  |     $9$      |       $9$       |  $Верно$   |
|  $7$  |     $11$     |       $11$       |  $Верно$   |
|  $8$  |     $11$     |       $11$       |  $Верно$   |
|  $9$  |     $13$     |       $13$       |  $Верно$   |
|  $10$ |     $14$     |       $14$       |  $Верно$   |
|  $11$ |     $16$     |       $16$       |  $Верно$   |
|  $12$ |     $16$     |       $16$       |  $Верно$   |

### Вывод

Таким образом, проверена гипотеза Шольца-Брауэра. Установлено, что она корректна для $n \in [1,12]$. Использовано одно из следствий из теоретических положений, чтобы ускорить проверку.

## Выводы

В ходе работы были изучены и применены на практике теоретические сведения об аддитивных цепочках и об алгоритмах нахождения аддитивных цепочек для заданного числа. Были реализованы алгоритм Яо и алгоритм дробления вектора индексов, который работает медленнее Яо, но гарантированно выводит минимально возможную цепочку для заданного числа. Также в ходе работы была проверена на практике гипотеза Шольца-Брауэра для $n \le 12$.